File: app/src/main/java/br/com/vicentec12/mygames/data/source/game/GameLocalDataSource.java
package br.com.vicentec12.mygames.data.source.game;

import androidx.annotation.NonNull;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import br.com.vicentec12.mygames.R;
import br.com.vicentec12.mygames.data.model.Console;
import br.com.vicentec12.mygames.data.model.Game;
import br.com.vicentec12.mygames.interfaces.Callbacks;
import br.com.vicentec12.mygames.util.AppExecutors;

@Singleton
public class GameLocalDataSource implements GameDataSource {

    public static final int ORDER_BY_NAME = 0;
    public static final int ORDER_BY_YEAR = 1;

    private final GameDao mGameDao;
    private final AppExecutors mAppExecutors;

    @Inject
    public GameLocalDataSource(@NonNull GameDao mGameDao, @NonNull AppExecutors mAppExecutors) {
        this.mGameDao = mGameDao;
        this.mAppExecutors = mAppExecutors;
    }

    @Override
    public void list(@NonNull Console console, int sortBy, OnGamesListedCallback callback) {
        mAppExecutors.diskIO().execute(() -> {
            List<Game> games;
            if (sortBy == ORDER_BY_NAME)
                games = mGameDao.list(console.getId());
            else
                games = mGameDao.listByYear(console.getId());
            mAppExecutors.mainThread().execute(() -> {
                if (games.isEmpty())
                    callback.onFailure();
                else
                    callback.onSuccess(games);
            });
        });
    }

    @Override
    public void get(int id, OnGameGetedCallback callback) {
        mAppExecutors.diskIO().execute(() -> {
            Game game = mGameDao.get(id);
            mAppExecutors.mainThread().execute(() -> {
                if (game == null)
                    callback.onFailure();
                else
                    callback.onSuccess(game);
            });
        });
    }

    @Override
    public void insert(@NonNull Game game, Callbacks.OnLocalCallback callback) {
        mAppExecutors.diskIO().execute(() -> {
            long rowIds = mGameDao.insert(game);
            mAppExecutors.mainThread().execute(() -> {
                if (rowIds > 0)
                    callback.onSuccess(R.string.message_game_inserted);
                else
                    callback.onFailure(R.string.error_message_game_insert);
            });
        });
    }

    @Override
    public void update(@NonNull Game game, Callbacks.OnLocalCallback callback) {
        mAppExecutors.diskIO().execute(() -> {
            int numUpdated = mGameDao.update(game);
            mAppExecutors.mainThread().execute(() -> {
                if (numUpdated > 0)
                    callback.onSuccess(R.string.message_game_updated);
                else
                    callback.onFailure(R.string.error_message_game_update);
            });
        });
    }

    @Override
    public void delete(@NonNull List<Game> games, OnGameDeletedCallback callback) {
        mAppExecutors.diskIO().execute(() -> {
            int numDeleted = mGameDao.delete(games);
            mAppExecutors.mainThread().execute(() -> {
                if (numDeleted > 0)
                    callback.onSuccess(R.plurals.plural_message_games_deleted, numDeleted);
                else
                    callback.onFailure(R.string.error_message_game_delete);
            });
        });
    }

    @Override
    public void deleteAll() {
        mAppExecutors.diskIO().execute(mGameDao::deleteAll);
    }

}

File: app/src/main/java/br/com/vicentec12/mygames/data/source/game/GameLocalDataSource.kt
package br.com.vicentec12.mygames.data.source.game

import br.com.vicentec12.mygames.R
import br.com.vicentec12.mygames.data.model.Game
import br.com.vicentec12.mygames.data.model.Game.Companion.COLUMN_NAME
import br.com.vicentec12.mygames.data.source.Result
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class GameLocalDataSource @Inject constructor(
        private val gameDao: GameDao
) : GameDataSource {

    override suspend fun list(mConsoleId: Long, mOrderBy: Int) = withContext(Dispatchers.IO) {
        val mGames = if (mOrderBy == COLUMN_NAME) gameDao.list(mConsoleId) else gameDao.listByYear(mConsoleId)
        if (mGames.isNullOrEmpty())
            Result.Error(0)
        else
            Result.Success(mGames)
    }

    override suspend fun get(mId: Long): Result<Game> = withContext(Dispatchers.IO) {
        val mGame = gameDao.get(mId)
        if (mGame != null)
            Result.Success(mGame)
        else
            Result.Error(0)
    }

    override suspend fun insert(mGame: Game) = withContext(Dispatchers.IO) {
        val mId = gameDao.insert(mGame)
        if (mId > 0)
            Result.Success(message = R.string.message_game_inserted)
        else
            Result.Error(R.string.error_message_game_insert)
    }

    override suspend fun update(mGame: Game) = withContext(Dispatchers.IO) {
        val mId = gameDao.update(mGame)
        if (mId > 0)
            Result.Success(message = R.string.message_game_updated)
        else
            Result.Error(R.string.error_message_game_update)
    }

    override suspend fun delete(mGames: List<Game>) = withContext(Dispatchers.IO) {
        val mNumDeleted = gameDao.delete(mGames)
        if (mNumDeleted > 0) {
            Result.Success(mNumDeleted, R.plurals.plural_message_games_deleted)
        } else
            Result.Error(R.string.error_message_game_delete)
    }

    override suspend fun deleteAll() = withContext(Dispatchers.IO) {
        gameDao.deleteAll()
        Result.Success<Nothing>(message = R.string.message_game_deleted)
    }

}